main/src/ca/uwaterloo/flix/language/phase/Desugar.scala

```diff
@@ -955,4 +959,85 @@ object Desugar {
       val p = pat.map(visitPattern)
       DesugaredAst.Pattern.Record.RecordLabelPattern(label, p, loc)
   }
+
+  /**
+    * Rewrites a `ForA` loop into a series of `Applicative.ap` calls.
+    * {{{
+    *   forA (
+    *         x <- xs;
+    *         y <- ys
+    *     ) yield exp
+    *
+    * }}}
+    * desugars to
+    * {{{
+    * Applicative.ap(Functor.map(x -> y -> exp, xs), ys)
+    * }}}
+    *
+    */
+  private def desugarApplicativeFor(frags: List[ForFragment.Generator], exp: WeededAst.Expr, loc: SourceLocation)(implicit flix: Flix): DesugaredAst.Expr = {
+    val fqnAp = "Applicative.ap"
+    val fqnMap = "Functor.map"
+    val yieldExp = visitExp(exp)
+
+    // Make lambda for Functor.map(lambda, ...). This lambda uses all patterns from the for-fragments.
+    val lambda = frags.foldRight(yieldExp) {
+      case (WeededAst.ForFragment.Generator(pat, _, loc1), acc) =>
+        val p = visitPattern(pat)
+        mkLambdaMatch(p, acc, loc1)
+    }
+
+    // Apply first fragment to Functor.map
+    val xs = visitExp(frags.head.exp)
```

Comment:
I assume we've already checked that frags is nonempty?

Reply:
Would be nice if we had Nel in our library...

Reply:
Empty frags is a Weeder error.